service/controller/resource/nodepool/create_deployment_uninitialized.go

```diff
@@ -0,0 +1,221 @@
+	if err != nil {
+		return currentState, microerror.Mask(err)
+	}
+
+	if machinePool == nil {
+		return currentState, microerror.Mask(ownerReferenceNotSet)
+	}
+
+	cluster, err := util.GetClusterFromMetadata(ctx, r.CtrlClient, azureMachinePool.ObjectMeta)
+	if err != nil {
+		return currentState, microerror.Mask(err)
+	}
+
+	credentialSecret, err := r.getCredentialSecret(ctx, *cluster)
+	if err != nil {
+		return currentState, microerror.Mask(err)
+	}
+
+	azureCluster, err := r.getAzureClusterFromCluster(ctx, cluster)
+	if err != nil {
+		return currentState, microerror.Mask(err)
+	}
+
+	release, err := r.getReleaseFromMetadata(ctx, azureMachinePool.ObjectMeta)
+	if err != nil {
+		return currentState, microerror.Mask(err)
+	}
+
+	deploymentsClient, err := r.ClientFactory.GetDeploymentsClient(credentialSecret.Namespace, credentialSecret.Name)
+	if err != nil {
+		return currentState, microerror.Mask(err)
+	}
+
+	storageAccountsClient, err := r.ClientFactory.GetStorageAccountsClient(credentialSecret.Namespace, credentialSecret.Name)
+	if err != nil {
+		return currentState, microerror.Mask(err)
+	}
+
+	virtualMachineScaleSetVMsClient, err := r.ClientFactory.GetVirtualMachineScaleSetVMsClient(credentialSecret.Namespace, credentialSecret.Name)
+	if err != nil {
+		return currentState, microerror.Mask(err)
+	}
+
+	desiredDeployment, err := r.getDesiredDeployment(ctx, storageAccountsClient, release, azureCluster, machinePool, &azureMachinePool)
+	if IsNotFound(err) {
+		r.Logger.LogCtx(ctx, "level", "debug", "message", "Azure resource not found, canceling resource")
+		return currentState, nil
+	} else if err != nil {
+		r.Logger.LogCtx(ctx, "level", "debug", "message", "canceling resource")
+		return currentState, microerror.Mask(err)
+	}
+
+	currentDeployment, err := deploymentsClient.Get(ctx, key.ClusterID(&azureMachinePool), key.NodePoolDeploymentName(&azureMachinePool))
+	if IsDeploymentNotFound(err) {
+		// We haven't created the deployment just yet, it's fine.
+	} else if err != nil {
+		return currentState, microerror.Mask(err)
+	}
+
+	deploymentIsOutOfDate, nodesNeedToBeRolled, err := r.deploymentIsOutOfDate(ctx, currentDeployment, desiredDeployment)
+	if err != nil {
+		return currentState, microerror.Mask(err)
+	}
+
+	if deploymentIsOutOfDate {
+		r.Logger.LogCtx(ctx, "level", "debug", "message", "template or parameters changed")
+
+		_, err = r.ensureDeployment(ctx, deploymentsClient, desiredDeployment, &azureMachinePool)
+		if err != nil {
+			r.Logger.LogCtx(ctx, "level", "debug", "message", "canceling resource")
+			return currentState, microerror.Mask(err)
+		}
+
+		if nodesNeedToBeRolled {
+			return ScaleUpWorkerVMSS, nil
+		}
+
+		return currentState, nil
+	}
+
+	// Start watcher on the instances to avoid stuck VMs to block the deployment progress forever
+	r.InstanceWatchdog.DeleteFailedVMSS(ctx, virtualMachineScaleSetVMsClient, key.ClusterID(&azureMachinePool), key.NodePoolVMSSName(&azureMachinePool))
```

Comment:
We need to monitor this a bit. I'm not fully sure how safe this is with node pools (since there can be hundreds of VMSSs per installation).